Add Configurable Tests Folder for `ape test` (#2576)

shwetd19 opened the pull request:
**Description**:
This PR addresses issue #2576 by implementing the following:
1. Allows specifying the `tests_folder` in `ape-config.yaml`.
2. Automatically uses the `tests/` folder if it exists.
3. Prevents running tests from unintended directories like `lib/`.

**fixes**: #2576

**What I did**:
- Added `tests_folder` to the `TestConfig` model in `ape/pytest/config.py`.
- Updated the test runner in `ape/pytest/runner.py` to prioritize the configured `tests_folder` or default to `tests/` if it exists.
- Modified the `ape test` CLI command to use the configured folder when no paths are provided.
- Added documentation for the new `tests_folder` option in `docs/userguides/config.md`.

**How I did it**:
- Extended the Pydantic configuration model to include `tests_folder` with a default of `tests`.
- Adjusted `PytestApeRunner` to check for the existence of `tests/` and use it over other directories, ensuring pytest only runs tests from the specified or default folder.
- Ensured compatibility with existing test workflows and validated folder existence to avoid errors.

**How to verify it**:
1. Create an `ape-config.yaml` with `test.tests_folder: custom_tests`.
2. Add a test file in `custom_tests/` and run `ape test` to confirm it executes only from `custom_tests/`.
3. Rename `custom_tests/` to `tests/` and run `ape test` to verify it defaults to `tests/`.
4. Run tests with a `lib/` folder containing test-like files to ensure they are ignored.
5. Check `pytest` output to confirm only the specified folder is used.

**Checklist**:
- [x] All changes are completed
- [x] Change is covered in tests (`tests/test_config.py`, `tests/test_pytest_runner.py`)
- [x] Documentation is complete (`docs/userguides/config.md`)

---

This PR enhances the `ape test` command’s usability by making the test folder configurable and prioritizing `tests/`, resolving the issue efficiently.

fubuloubu commented on docs/userguides/config.md:
````diff
@@ -384,14 +384,22 @@ To learn more about how request headers work in Ape, see [this section of the Ne
 ## Testing
 
-Configure your test accounts:
+Configure your test accounts and test folder:
 
 ```yaml
 test:
   mnemonic: test test test test test test test test test test test junk
   number_of_accounts: 5
+
+# Optional: Configure the tests folder location
+tests_folder: tests
````
I feel like explaining the test folders should come after the test accounts, as a separate topic.

fubuloubu reviewed (commented):
Seems like a little work needed on the new test, but I do wonder how this works with #2513

shwetd19 commented:
Hey @fubuloubu I've Addressed your feedback by restructuring `docs/userguides/config.md` to separate test accounts and `tests_folder` sections, with accounts first. Fixed failing tests in `tests/test_pytest_runner.py` by adding a `setup_test_dir` fixture for robust directory creation in CI. I've also ensured compatibility with #2513 by keeping it's custom pytest args in `PytestApeRunner` and adding tests for multiple test directories.
